inotify_poller.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build linux
+
+package fsnotify
+
+import (
+	"errors"
+	"os"
+	"syscall"
+)
+
+type fdPoller struct {
+	fd   int    // File descriptor (as returned by the inotify_init() syscall)
+	epfd int    // Epoll file descriptor
+	pipe [2]int // Pipe for waking up
+}
+
+// Create a new inotify poller.
+// This creates an inotify handler, and an epoll handler.
+func newFdPoller() (*fdPoller, error) {
+	var errno error
+	poller := new(fdPoller)
+
+	// Create inotify fd
+	poller.fd, errno = syscall.InotifyInit()
```

Comment:
Likewise, NewWatcher should be responsible for closing the inotify fd if poller has an error.

Poller could in theory work for any fd passed in and only clean up its own resources.